### PR TITLE
support for TCP tunnels - tests

### PR DESCRIPTION
**Pull request by rjeczalik**
/cc @cihangir

### CODE REVIEW

**Comment by cihangir** [comment]
cc @fatih

**Comment by rjeczalik** [comment]
If @cihangir is calling for backup then I'm in trouble.

pic

File: tunneltest/tunneltest.go [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+	"os"
+	"sort"
+	"strconv"
+	"sync"
+	"testing"
+
+	"github.com/koding/tunnel"
+)
+
+var noDebug = os.Getenv("NO_DEBUG") == "1"
```
Comment (cihangir) [inline tunneltest/tunneltest.go]: i think this is not required, you are already utilizing testing.Verbose()

File: helper_test.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+			Type:      tunneltest.TypeHTTP,
+			LocalAddr: "127.0.0.1:0",
+			Handler:   handler,
+		},
+	}
+}
+
+func singleTCP(handler interface{}) map[string]*tunneltest.Tunnel {
+	return map[string]*tunneltest.Tunnel{
+		"http": {
```
Comment (cihangir) [inline helper_test.go]: you dont need http here

File: helper_test.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+func handlerEchoTCP(conn net.Conn) {
+	io.Copy(conn, conn)
+}
+
+func handlerLatencyEchoTCP(conn net.Conn) {
+	sleep()
+	handlerEchoTCP(conn)
+}
```
Comment (cihangir) [inline helper_test.go]: :+1:

File: helper_test.go (hunk 2 above)
Comment (rjeczalik) [inline helper_test.go]: it's just identifier for the tunnel, it can be any non-empty string

File: tunneltest/tunneltest.go [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+}
+
+func (tt *TunnelTest) Serve(tunnels map[string]*Tunnel) error {
+	if len(tunnels) == 0 {
+		return errors.New("no tunnels to serve")
+	}
+
+	// Since one tunnels depends on others do 3 passes to start them
+	// all, each started tunnel is removed from the tunnels map.
+	// After 3 passes all of them must be started, otherwise the
```
Comment (cihangir) [inline tunneltest/tunneltest.go]: in my opinion removing items from a map in a `serveDeps` named  function is so implicit, can we change this naming?

File: helper_test.go (hunk 2 above)
Comment (cihangir) [inline helper_test.go]: strange thing is if i remove or change the name, TestSingleTCP and  TestMultipleTCP tests are failing with  ``` unable to start tunnels due to unresolved dependencies: [ tcp]```

File: tunneltest/tunneltest.go [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+	}
+
+	if err = tt.Serve(tunnels); err != nil {
+		return nil, err
+	}
+
+	return tt, nil
+}
+
+func (tt *TunnelTest) serveSingle(ident string, t *Tunnel) (bool, error) {
```
Comment (cihangir) [inline tunneltest/tunneltest.go]: you can refactor this function later, it has 150 lines :)

Comment (rjeczalik) [inline tunneltest/tunneltest.go]: definitely, added a TODO

File: tunneltest/tunneltest.go (hunk 4 above)
Comment (rjeczalik) [inline tunneltest/tunneltest.go]: done -> `popServedDeps`

File: helper_test.go (hunk 2 above)
Comment (rjeczalik) [inline helper_test.go]: the tunnel with ident `tcp` depends on already existing tunnel client with ident `http`, that's why renaming `http` to something else makes tunneltest fail when verifying deps - updated doc so it's more clear

File: tunneltest/tunneltest.go (hunk 1 above)
Comment (rjeczalik) [inline tunneltest/tunneltest.go]: done, removed

**Comment by cihangir** [comment]
:+1:

**Pull request merged**